src/main/java/software/amazon/smithy/lsp/language/FoldingRangeHandler.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.lsp.language;
+
+import org.eclipse.lsp4j.FoldingRange;
+import org.eclipse.lsp4j.Range;
+import software.amazon.smithy.lsp.document.Document;
+import software.amazon.smithy.lsp.document.DocumentImports;
+import software.amazon.smithy.lsp.syntax.Syntax;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+
+
+public record FoldingRangeHandler(Document document, DocumentImports documentImports,
+                                  List<Syntax.Statement> statements) {
+    /**
+     * Main public handle function in the handler class
+     *
+     * @return A list of FoldingRange
+     */
+    public List<FoldingRange> handle() {
+        return generateFoldingRanges();
+
+    }
+
+    /**
+     * Check if the given range can be folded
+     *
+     * @param range The Range object held the start line and end line info for the folding range
+     * @return If the given Range can be folded (end line - start line >= 1)
+     */
+    private boolean isFoldable(Range range) {
+        return range != null && range.getEnd().getLine() - range.getStart().getLine() >= 1;
+    }
+
+    /**
+     * Create a fold range object with given range and add it to the foldingRanges list
+     *
+     * @param range         The Range object held the start line and end line info for the folding range
+     * @param foldingRanges The list holds folding ranges for current document
+     */
+    private void addFoldingRange(Range range, List<FoldingRange> foldingRanges) {
```

Comment:
Another suggestion - you can avoid allocating the intermediate `Range` by just using `Document#lineOfIndex` for the start and end. That method can return -1 if the line doesn't exist (basically equivalent to the range null check).